src/MiddlewareDispatcher.php

```diff
@@ -13,88 +13,89 @@
 /**
  * MiddlewareDispatcher
  */
-final class MiddlewareDispatcher implements RequestHandlerInterface, MiddlewareInterface
+final class MiddlewareDispatcher implements MiddlewareInterface
 {
     /**
      * @var MiddlewareInterface[]
      */
-    private $middlewares = [];
+    private array $middlewares = [];
 
-    /**
-     * @var RequestHandlerInterface|null
-     */
-    private $nextHandler;
+    private RequestHandlerInterface $nextHandler;
+    private ContainerInterface $container;
 
     /**
-     * @var ContainerInterface
+     * Contains a chain of middleware wrapped in handlers.
+     * Each handler points to the handler of middleware that will be processed next.
+     * @var RequestHandlerInterface|null stack of middleware
      */
-    private $container;
+    private ?RequestHandlerInterface $stack = null;
 
     public function __construct(
-        array $middlewares,
         ContainerInterface $container,
         RequestHandlerInterface $nextHandler = null
     ) {
-        if ($middlewares === []) {
-            throw new \InvalidArgumentException('Middlewares should be defined.');
-        }
-
         $this->container = $container;
 
-        foreach ($middlewares as $middleware) {
-            $this->add($middleware);
-        }
-
         $responseFactory = $container->get(ResponseFactoryInterface::class);
 
         $this->nextHandler = $nextHandler ?? new NotFoundHandler($responseFactory);
     }
 
-    private function addCallable(callable $callback): void
-    {
-        $this->middlewares[] = new Callback($callback, $this->container);
-    }
-
-    public function add($middleware): void
+    /**
+     * @param callable|MiddlewareInterface $middleware
+     * @return self
+     */
+    public function addMiddleware($middleware): self
     {
         if (is_callable($middleware)) {
-            $this->addCallable($middleware);
-        } elseif ($middleware instanceof MiddlewareInterface) {
-            $this->middlewares[] = $middleware;
-        } else {
+            $middleware = new Callback($middleware, $this->container);
+        }
+
+        if (!$middleware instanceof MiddlewareInterface) {
             throw new \InvalidArgumentException('Middleware should be either callable or MiddlewareInterface instance. ' . get_class($middleware) . ' given.');
         }
+
+        array_unshift($this->middlewares, $middleware);
```

Comment:
```suggestion
        $this->middlewares[] = $middleware;
```

Reply:
I thought exactly the same but I so lazy...